### ISSUE TITLE
*(basic_string::end()) must NOT be UB

### ISSUE THREAD

**Issue by kelbon** [microsoft/STL#4346]
Standard guarantees that [.data(); data() + size()] is valid range.
Standard guarantees, that operator[](size()) is valid == '\0'
gcc and clang check for <= size() in assert for *string.end();

But msvc aborted on *std::string().end()

It is incorrect, please fix

**Comment by sylveon** [comment]
While the standard does say that `[data(), data() + size()]`, it only says that `end()` returns "[An iterator which is the past-the-end value.](https://eel.is/c++draft/basic.string#string.iterators-2)".

Past the end iterators are always invalid to dereference, so this is correct behavior.

**Comment by kelbon** [comment]
> While the standard does say that `[data(), data() + size()]`, it only says that `end()` returns "[An iterator which is the past-the-end value.](https://eel.is/c++draft/basic.string#string.iterators-2)".
> 
> Past the end iterators are always invalid to dereference, so this is correct behavior.

There are no such word in standard. Standard only says, that standard library (algorithm) cannot assume, that sentinel iterator is dereferensable 

https://eel.is/c++draft/iterator.requirements#general-7.sentence-4

**Comment by kelbon** [comment]
> While the standard does say that `[data(), data() + size()]`, it only says that `end()` returns "[An iterator which is the past-the-end value.](https://eel.is/c++draft/basic.string#string.iterators-2)".
> 
> Past the end iterators are always invalid to dereference, so this is correct behavior.

If it is undefined behavior, why consteval function do not catch it? Then fix 'consteval' https://godbolt.org/z/4Y5c3TK16

(localy compilation failure because of incorrect check in stdlib)

**Comment by frederick-vs-ja** [comment]
> Standard guarantees that [.data(); data() + size()] is valid range.
> Standard guarantees, that operator [ ] (size()) is valid == '\0'

Right. These require that `data() + size()` is dereferenceable. However, there doesn't seem guarantee that `end()` is also dereferenceable.

Note that [[string.access]/2](https://eel.is/c++draft/string.access#2) says:
> _Returns_: `*(begin() + pos)` if `pos < size()`. Otherwise, returns a reference to an object of type `charT` with value `charT()`, where modifying the object to any value other than `charT()` leads to undefined behavior.

If `end()` were already dereferenceable, we can directly say `pos <= size()` instead.

I think it may need an LWG issue to add the guarantee you wanted.

> If it is undefined behavior, why consteval function do not catch it? Then fix 'consteval' https://godbolt.org/z/4Y5c3TK16

UB additionally specified in for the standard library is generally not required to be diagnosed, see [[expr.const]/6.1](https://eel.is/c++draft/expr.const#6.1):
> It is unspecified whether _E_ is a core constant expression if _E_ satisfies the constraints of a core constant expression, but evaluation of _E_ would evaluate
> (6.1) - an operation that has undefined behavior as specified in [[library]](https://eel.is/c++draft/library) through [[thread]](https://eel.is/c++draft/thread),

**Comment by TheStormN** [comment]
It's interesting topic as the std::string does not count the terminating zero as part of the size, then I guess end() returning that same terminating zero should be fine?

**Comment by StephanTLavavej** [comment]
Thanks for reporting this issue. We're resolving this as invalid - as @sylveon mentioned above, dereferencing end iterators is undefined behavior according to the Standard, even if a null terminator is physically present there, and MSVC's STL is deliberately being strict about enforcing this in debug mode.

(The rationale is essentially that the null terminator is present to make `.c_str()` work, for backwards compatibility with C APIs. However, the `basic_string` really "wants" to present itself as storing a sequence of characters without appending a null terminator, so only the range `[begin, end)` can be obtained by dereferencing its iterators.)
